### PR TITLE
GPII-3553: Added browser coverage to overall report.

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [GPII-3553](https://issues.gpii.net/browse/GPII-3553) for details.

### CODE REVIEW

File: tests/js/lib/instrument-source.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+fluid.registerNamespace("gpii.tests.handlebars.instrumentation");
+
+gpii.tests.handlebars.instrumentation.options = {
+    sources:    ["./src/**/*.js", "./index.js"],
+    excludes:   [],
+    nonSources: [
+        "./tests/**",
+        "./Gruntfile.js",
+        "./node_modules/gpii-binder/src/js/binder.js",
```
Comment (amb26) [inline tests/js/lib/instrument-source.js:23]: Why does anything from within our top-level node_modules explicitly need to be excluded?

**Comment by amb26** [comment]
Tests fail on node 10 - 
````
10:11:44.732:  jq: FAIL: Module "Unit tests for i18n functions." Test name "Large writes should be complete before we are notified of a file
 change..." - Message: Unexpected failure in test case (see following log for more details): Callback must be a function
10:11:44.733:  jq: Source:     at kettle.test.handleUncaughtException (E:\source\gits\gpii-handlebars\node_modules\kettle\lib\test\KettleTes
tUtils.js:31:15)
    at Object.fire (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1568:40)
    at process.onUncaughtException (E:\source\gits\gpii-handlebars\node_modules\infusion\src\module\fluid.js:229:31)
    at process.emit (events.js:182:13)
    at processEmit [as emit] (E:\source\gits\gpii-handlebars\node_modules\nyc\node_modules\signal-exit\index.js:155:32)
    at process._fatalException (internal/bootstrap/node.js:491:27)
10:11:44.735:  jq: Test concluded - Module "Unit tests for i18n functions." Test name "Large writes should be complete before we are notifie
d of a file change...": 3/4 passed - FAIL
````
via
````
10:11:44.649:  FATAL ERROR: Uncaught exception: Callback must be a function
TypeError [ERR_INVALID_CALLBACK]: Callback must be a function
    at makeCallback (fs.js:137:11)
    at Object.unlink (fs.js:936:14)
    at E:\source\gits\gpii-handlebars\tests\js\server\watcher-tests.js:13:688
````

I got bitten by this recently as well, they have tightened up the requirements on calling fs.unlink -> fs.unlinkSync

**Comment by the-t-in-rtf** [comment]
I replaced the sole mistaken usage of fs.unlink in the tests and updated all dependencies.  @amb26, this should be ready for further review.

File: tests/js/lib/instrument-source.js (hunk 1 above)
Comment (the-t-in-rtf) [reply tests/js/lib/instrument-source.js:23]: That's not an exclusion.  Each entry is a "non source" file that should not be instrumented, but that still needs to be served up by the express instance we use with Testem.

**Comment by amb26** [comment]
Cheers, this now works fine - just a note about some low spots of coverage - 
i) hasRequiredOptions.js has only 25% branch coverage since all of the pathway for reporting errors in checkRequiredOptions is uncovered

ii) The entirety of gpii.handlebars.helper.initBlock.generateInitBlock is uncovered - file has 0% branch coverage

iii) some of the tests themselves have been selected for coverage given we get reporting for gpii-handlebars/tests/js/server and lib

**Comment by the-t-in-rtf** [comment]
So, the test content in the report is easy, I added a simple exclude that I'll commit shortly.  I'd need to write new tests for the required options failure cases, but that's certainly par for the course.

 What's harder is point 2.  The initBlock helper is exercised as part of the browser tests, but nothing is collecting that coverage data.  The way we run the browser tests is to instrument the whole directory and then run the webdriver tests from that directory.  We can't simply use nyc with already-instrumented code.  If we avoid instrumenting the "server" code using the "non source" option we discussed earlier, we can  run the browser tests using nyc.  This gives us a partial restoration of coverage data, including the initBlock coverage reporting.  The problem comes with the "common" code, i.e. code used in both the browser and server.  That has to be instrumented to get coverage in the browser, but can't be instrumented if we expect nyc to be able to collect coverage data.  If we're willing to only get that coverage from the browser side, that change is more or less already ready.

If we really want that last drop of coverage for the "common" fixtures hit from the server side,  a medium-term option might be to run the browser tests twice, once to get the browser coverage, and once to get the server-side fixture coverage.  This would require refactoring the tests, as they currently explicitly wait for coverage data to be sent back.  I'd have to think of how to do this without making two parallel sets of browser test fixtures, but that's probably an hour or two of work.

The long-term way to handle this is to develop and explicitly test a means of collecting coverage simultaneously for the browser fixtures and also for the server-side fixtures exercised by the browser.  I suspected we would eventually get there, but we dodged this particular bullet with universal.  If I was going to address the use case, it makes the most sense to write it up as an improvement to gpii-testem and test that work with an update of gpii-handlebars.  That would be a few days' work including the review.

For this pull, I'd argue that we should go for the first option with a few follow-up steps, namely:

1. Configure the the browser tests to run using nyc.
2. Write up the feature request against gpii-testem.
3. Leave a clear TODO in this package linking to the ticket.
4. Agree to revisit if we have another project that needs the feature.

**Comment by the-t-in-rtf** [comment]
Another option is to make it safe to request the browser fixtures from pure node tests, which is similar to my "medium term" suggestion above in terms of the effort involved.

**Comment by the-t-in-rtf** [comment]
Antranig and I discussed this and agreed to document the required new feature and live with what we can do now.  I did try running the existing browser tests using nyc, although this does restore the initBlock coverage, it clobbers the reporting on "common" fixtures with a nonsense report about the instrumented code.  The status quo is simpler and I have left that until the new feature is available.

**Comment by the-t-in-rtf** [comment]
@amb26, we have as much coverage data as we can easily collect now, including new tests for the "check required options" grade that will eventually be replaced by gpii-json-schema.  This should be ready for further review.

**Comment by the-t-in-rtf** [comment]
Thanks, @amb26.  I have cut the full release and updated the forward-facing version in master.

**Pull request merged**